Implement base routing functionality.

mwistrand opened the pull request:
One can get a sense for how this is designed to work by looking at `tests/functional/routerApp/Router.ts`.
1. `Router` is the central hub for routing, and is responsible for calling the various lifecycle methods on individual routes.
2. For interacting with the environment (e.g., responding to `popstate` or `hashchange` events), `Router` delegates to `RouterSource` objects. Three `RouterSource` classes are provided: `HtmlHistorySource` for using the History/`pushState` API, `HashSource` for using `window.location.hash`, and `DefaultBrowserSource` for allowing the router to work with hard page refreshes. While it is possible to specify which source to use when instantiating `Router`, by default `HtmlHistorySource` is used where support is available, with `HashSource` used as the fallback.
3. `Route` and `DefaultRoute` are responsible for handling individual paths. While `Route` is designed to handle a specific path, `DefaultRoute` will respond to any path.
4. Routes use lifecycle methods (`beforeExit`, `beforeEnter`, `enter`, `exit`, and `change`) to manage flow. Route changes can be canceled by calling `event.preventDefault()` from within `beforeExit` or `beforeEnter`. Additionally, `Router` emits `beforechange` and `change` events, the former also useful for canceling route changes (again, via `event.preventDefault()`).
5. Routes can be optionally grouped under `RouteGroup` objects, which more-or-less behave as mediators between the `Router`'s path and the paths of the individual routes. This allows routes to be used in multiple places without concern for the parent path.

Commit:
Implement routing functionality.

* Add Router and router "sources" for managing the HTML History API and `hashchange` events.
* Provide RouteGroup for optionally grouping routes under a sub-path.
* Add Route and DefaultRoute classes for responding to individual paths.
* Add PathRule for converting paths into more useful segments.
* Add unit and functional tests.

Commit:
Add dojo-core type definition.

mwistrand commented:
As of now, `Router` allows for an optional `error` callback for error handling. If no such callback is provided, and the error is not a `CancelNavigationError` object, then the error is re-thrown. But now I'm wondering if emitting an `error` event would be better, since events are already being emitted by `Router`.

mwistrand commented on src/Router.ts:
```diff
@@ -0,0 +1,186 @@
+		this.routes = this._rule = this.source = null;
+	}
+
+	emit(data: EventObject): void {
```
This method perhaps should be removed, since emitting events is something handled by `Router` itself and not other objects.

kitsonk commented on src/DefaultBrowserSource.ts:
```diff
@@ -0,0 +1,35 @@
+	}
+
+	go(path: string = null): void {
+		if (path === null || !this._current) {
```
So if I do `.go('mypath')` but I don't have a `_current`, I won't actually navigate, right?

edhager replied:
This logic is confusing.  I don't see any tests for this module.  Am I missing something?

mwistrand replied:
There are missing tests for this; I'll add those ASAP. The point of the `RouterSource` objects is to make it possible to have a single `Router` class that has the ability to interact with different environments without having to know anything about them. If you look at the implementations/tests for `HtmlHistorySource`, hopefully the interface is less confusing. This particular module is just weird because the logic required to make it work is so minimal.

mwistrand replied:
@kitsonk You're right; I'm going to have to think through a better implementation. The idea here is that if `Router` calls `RouterSource#go` without a path, then the `RouterSource` will get the default path from the enviroment and emit an event that will trigger `Router#go`. In the manual tests I've done this works fine, but as you pointed out this is brittle. I have a working implementation that's more solid, and I'll update this PR once the tests are in place.

kitsonk commented on src/DefaultBrowserSource.ts:
```diff
@@ -0,0 +1,35 @@
+		}
+		else if (path !== this._current) {
+			global.location = global.location.origin + path;
+		}
```
I know this is an edge case, but if I do `.go(undefined)`, it isn't type guarded by TS, but with the add operator, it gets coerced into a string and `global.location` would end up being a string value of `'undefined'`.

On second thought, it isn't an edge case... consider the following:

``` typescript
let myPath: string;
let defaultBrowserSource = new DefaultBrowserSource();
defaultBrowserSource.go(myPath); /* global.location === 'undefined' */
```

mwistrand replied:
@kitsonk Did you generate this result in a browser? If I run pass `undefined` to `DefaultBrowserSource#go`, the default value is used (`null`) and the first block is entered.

kitsonk commented on src/DefaultRoute.ts:
```diff
@@ -0,0 +1,38 @@
+export default class DefaultRoute implements MatchableRoute {
+	protected _rule: PathRule;
+
+	beforeEnter: (kwArgs: CancelableNavigationArgs) => (Promise<void> | void);
```
I thought we were going to try to drop `kwArgs` in lieu of `options`, @kfranqueiro?  Maybe we should add it to the style guide?

kfranqueiro replied:
Maybe... (I have one reservation about that rename, see https://github.com/dojo/core/issues/28)

kitsonk commented:
Why isn't this package heavily using `dojo-core/Registry`?

kitsonk commented on src/HashSource.ts:
```diff
@@ -0,0 +1,84 @@
+
+	constructor(prefix: string = '!') {
+		if (!has('host-browser')) {
+			throw new Error('HashSource requires a browser.');
```
Does this code have to be so non-isomorphic?  Can't we narrow it down?  It is really looking for two features: `global.location` and `global.addEventListener`/`global.removeEventListener`.  If those are present why do we care if this is a browser or not?

mwistrand replied:
The `host-browser` test already checks for `global.location`, but I guess we don't have to care whether `global.document` exists as well.

kitsonk commented on src/has.ts:
```diff
@@ -0,0 +1,6 @@
+import global from 'dojo-core/global';
+import { add } from 'dojo-core/has';
+
+add('history', 'history' in global);
```
Maybe there is a better name for this?  Again, maybe a styleguide thing @kfranqueiro in that we should try to name features in what sort of "standard" introduced them.  The History API was in theory HTML5, so would `html5-history` be a better name?

Also, while it maybe overkill, a more authoritative test would be `global.window && global.window.history && 'pushState' in global.window.history`.

kfranqueiro replied:
+1 to `html5-history`.  Yeah I guess the style guide could use a couple of updates.

mwistrand replied:
I agree that `html5-history` is a better name. I didn't see `html5` prefixes elsewhere, so I left it off. But I certainly wouldn't mind adding it.

kfranqueiro replied:
The most consistent example of prefixing we have so far is `dom-` in core and dom.  It's hard for me to figure a clear and concise hard and fast rule for this though, since we also have several unprefixed things in core, e.g. xhr/xhr2, formdata, postmessage, raf...  Not sure if any of those would also warrant prefixes.

kitsonk replied:
I have tried to deduce the standard that introduced them... `es5`, `es6`, `html4` (for addEventListener, though that is arguable that is just `dom`).  Since we are only supporting IE9+ and other modern browsers, then I guess the set of prefixes that should be covered off are `es5`, `es6`, `es7`, `html5`, `css3`, `css2` and `dom`.  Can't think of anything else...

dylans replied:
addEventListener is technically from dom level 2 events :)  HTML4 started out with what are known as DOM level 0 events only (onclick="", etc.). So I'd probably call that dom2.

kfranqueiro replied:
Honestly, for the dom ones particularly I could see leaving off the number because who the heck is going to remember them?

dylans replied:
Apparently only me. :)  I was thinking it would be useful to distinguish between something added later, such as MutationObserver which iirc is dom level4, but just calling them dom is probably sufficient, unless they are something that is changed across DOM versions (the only thing that comes to mind is deprecated mutation events).

kitsonk replied:
Yeah, I agree, it is only @dylans who is going to remember... :grinning:

mwistrand commented:
@kitsonk I wasn't sure how feasible `dojo-core/Registry` would be with the all of the sub-route handling, but an attempt at incorporating it is on my to-do list.

kitsonk commented:
@mwistrand it allows you to generate your own matching functions, so in theory, it should work and of course you could do registries within registries potentially for your sub-routing.

Commit:
Add "error" event to Router.

* Remove `Router#error`.
* Add "error" event.
* Add `errors/MissingRouteError`.

mwistrand commented on src/Router.ts:
```diff
@@ -0,0 +1,186 @@
+
+	canceled: NavigationArgs;
+	current: NavigationArgs;
+	error: (error: Error) => void;
```
If this were to be removed in a favor of an "error" event (which IMO it should), then the promise returned by `Router#go` should both emit this error _and_ reject, correct?

Commit:
`has` test cleanup.

* Rename "history" test to "html5-history".
* Replace `has('host-browser')` test in HashSource with a more specific test.

Commit:
Add functional tests for DefaultBrowserSource.

Commit:
Update routing to use dojo-core/Registry.

* Implement `PathRegistry` module.
* Replace `RouteManager` with `RouteGroup`.
* Rename `routing.ts` to `interfaces.ts`.
* Refactor unit tests, add `PathRegistry` tests.
* Add jsdoc comments.

mwistrand commented:
The routing package now provides a `PathRegistry` class that extends `dojo-core/Registry`, and is used to handle all route matching. This allowed me to completely replace `RouteManager` with `RouteGroup`, and also to more cleanly separate concerns across the different modules.

Other changes with this commit include:
- Added JSDoc comments.
- General interface cleanup.
- Remove `Router#run`, since all it did was call `Router#go` with its path set to either "/" or `Router#source.currentPath`. As `Router#source` is public, it seemed less confusing to remove this method entirely.
- Bug fix: `RouteGroup` path rules match the beginning of paths, while `Route` path rules match entire paths (previously `PathRule` only matched the beginning of paths).

mwistrand commented:
There are still certain aspects of the underlying API that I'm not 100% sold on. For example, the purpose of route groups is so individual route/group objects can be reused across different groups. However, consider the following scenario:

``` typescript
const commentsGroup = new RouteGroup({
    path: 'comments/',
    defaultRoute: new DefaultRoute({
        enter() {}
    })
});
const articlesGroup = new RouteGroup({
    path: 'articles/{id}/',
    routes: [
        commentsGroup,

        new Route({
            path: 'edit/',
            enter() {}
        })
    ],
    defaultRoute: new DefaultRoute({
        enter() {}
    })
});
commentsGroup.addRoute(new Route({
    path: '{id}/',
    enter()
}));
```

In the above, the new route is added to `commentsGroup`, but any path the user navigates to will be first passed through `articleGroup`, so the combined path template is "articles/{id}/comments/{id}/", with two identical tokens. One possible solution would be to completely lock down `RouteGroup` so that routes can only be added at instantiation. Another solution would be to keep a reference to all routers on all groups, which would enable cascading validation when adding new routes. However, neither of these solutions strike me as being particularly clean. Any suggestions here would be appreciated.

devpaul commented on src/Router.ts:
```diff
@@ -0,0 +1,198 @@
+
+if (has('host-browser')) {
+	defaultSource = new (<any> DefaultSource)();
+}
```
would you make `defaultSource` a const and set its value using a ternary?

mwistrand replied:
Since the ternary resolves to a value before it's assigned to `DefaultSource`, I believe the use of `const` is valid.

devpaul replied:
Yeppers. `const` allows a single assignment :)

mwistrand replied:
Oh, you're not asking whether the use of `const` for `DefaultSource` is valid; you're suggesting that the same could be done for `defaultSource`. Gotcha.

devpaul commented on src/interfaces.ts:
```diff
@@ -0,0 +1,59 @@
+	rule?: PathRule;
+}
+
+export interface NavigationArgs extends EventObject {
```
I feel we should rename this type. Arguments are typically only passed into methods; however, NagivationArgs are returned from PathRegistry#match() and altered in Router#go().

Maybe NavigationPath, although that might be a bit ambiguous it is being held by the PathRegistry. Another option could be NavigationDescriptor.

mwistrand commented:
I agree. I was actually considering renaming this `RouteData` a là Mayhem's router, but held off in favor of the proposal.

devpaul commented on src/Router.ts:
```diff
@@ -0,0 +1,198 @@
+				handle.destroy();
+				resolve();
+			});
+			object[method](event);
```
IIRC, after won't catch an exception even though the promise will.

mwistrand replied:
That is correct. Routing is promise-based, but since it incorporates events we also needed a way to determine from the `dojo-core/Evented` system whether the promise should be resolved. Since `dojo-core/Evented` is synchronous, using `after` makes it possible to determine whether the callbacks executed without error, and therefore whether the underlying promise ought to be resolved.

devpaul replied:
The approach is really good. Looking at the after callback again, I'm not sure I should be concerned.

My thought was if `object[method]()` throws then `handle.destroy()` won't be called which would keep a reference to the closure tied to the object and if `object[method]()` is called again then the attached after callback will be called on success.  I think that I'm a little paranoid because `_getEventPromise()` is very generic.

Anyway, I wanted to highlight that and if you're comfortable with it then this doesn't need to change.

devpaul commented on src/interfaces.ts:
```diff
@@ -0,0 +1,59 @@
+import { EventObject } from 'dojo-core/interfaces';
```
Should we rename `interfaces.ts` to `interfaces.d.ts`? I only see declarations here.

devpaul commented:
Regarding composed routes where you might have a route of `/group/{id}/ticket/{id}` composed of a parent-child route handler Route('group/{id}') -> Route('ticket/{id}') I would think that under normal circumstances that the implementor would be aware of the name collision and rename the route parameters accordingly if they needed access.

However, this brings up an interesting point. If I am composing routes in a parent-child system I may also want to compose their data. For instance Router('group/{id}) can have a handler hook to fetch a group and return it via a new Promise({ group: id}) and Router('ticket/{id}') can be provided this data in addition to route parameters.

Applying this thought to a system like Express with a router model, dojo-router could provide similar functionality using this paradigm by creating Router('') => Promise({ req: HTTPRequest, res: HTTPResult, preventPropigation: Boolean}) --> Router('group/{id}') => Promise({ req, res, prevent, group: Group}) -> etc....

We should investigate adding this functionality after completing this PR.

Commit:
Rename interfaces.ts to interfaces.d.ts

mwistrand commented on src/interfaces.d.ts:
```diff
@@ -0,0 +1,59 @@
+	rule?: PathRule;
+}
+
+export interface NavigationArgs extends EventObject {
```
@devpaul originally commented (before I renamed this file):

I feel we should rename this type. Arguments are typically only passed into methods; however, NagivationArgs are returned from PathRegistry#match() and altered in Router#go().

> Maybe NavigationPath, although that might be a bit ambiguous it is being held by the PathRegistry.
> Another option could be NavigationDescriptor."

mwistrand replied:
And in response to that comment, I considered changing this to `RouteData` but held off in favor of keeping the name from the proposal. Are there any other suggestions for an alternative name to `NavigationArgs`?

kitsonk replied:
Since it extends `Event`, shouldn't it be named something like `NavigationEvent` especially if it contains things like `target` etc...

mwistrand replied:
Actually, yes; `NavigationEvent` makes sense.

Commit:
Pass NavigationArgs to Route#enter.

Previously, `Route#enter` excepted a path string as its sole argument. However, this is inconsistent with the `Route#change` implementation, and much less robust.

Commit:
Update typescript and tslint versions.

Commit:
Interface name cleanup.

* Change `NavigationArgs` to `NavigationEvent`.
* Change `CancelableNavigationArgs` to `CancelableNavigationEvent`.
* Remove unused modules/variables.

les2 commented on src/DefaultRoute.ts:
```diff
@@ -0,0 +1,74 @@
+	/**
+	 * A required lifecycle method that is called when `this` route is entered.
+	 *
+	 * @param path The full, normalized path matched by the router.
```
doc: There is no formal parameter named `path`.

novemberborn commented:
Closing in favor of #8.
